### ISSUE TITLE
AzureServiceBusResource should include possibility to specify subscriptions for each topic.

### ISSUE THREAD

**Issue by bjorkstromm** [dotnet/aspire#1789]
Azure Service Bus Resource in `Aspire.Hosting.Azure` should include the possibility for users to specify subscriptions for each topic. Today, if Service Bus is used with `Aspire.Hosting.Azure.Provisioning`, and one or more topic is specified, they can't be used from a subscriber point-of-view since no subscriptions exist.

https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-queues-topics-subscriptions#topics-and-subscriptions

Basically this needs to be updated to include a list of subscriptions for each topic, and provisioner then needs to be updated accordingly in order to support creating subscriptions.
https://github.com/dotnet/aspire/blob/0d1f8ea7750b69c8f221816077e9e5a5baa1bc0a/src/Aspire.Hosting.Azure/AzureServiceBusResource.cs#L23-L26

**Comment by bjorkstromm** [comment]
@davidfowl, as this change might break manifest (depending on how it's implemented) it might be wise to consider a fix before GA.

If you think it's worth prioritizing, I can send a PR for a fix.

**Comment by davidfowl** [comment]
Yes we should fix this.

**Comment by bjorkstromm** [comment]
> Yes we should fix this.

We should probably look at creating separate resources for queues and topics while we are at it. These would probably be better off as parent-child relationships. Azure Service Bus Namespace being the parent, and queues and topics (with subscriptions) being children.

**Comment by davidfowl** [comment]
I agree, we should re-design this. AZD can support both formats for a short time frame but we can change this in preview4. Would you be willing to send a PR?

cc @ellismg for heads up.

**Comment by bjorkstromm** [comment]
> Would you be willing to send a PR?

Consider it done 🙂

**Comment by davidfowl** [comment]
Assigning to you @bjorkstromm !

**Comment by davidfowl** [comment]
@bjorkstromm just a heads up. I'm doing a massive refactoring of our azure resources and provisioner so maybe wait a little with this one.

https://github.com/dotnet/aspire/pull/2056

**Comment by bjorkstromm** [comment]
@davidfowl thanks for the heads up. I have the PR almost ready (just wanted to do some verifications before submitting).

Do you happen to have a branch for the massive refactoring? Just so I can have a look, and maybe try rebasing my work on top of that.

**Comment by davidfowl** [comment]
https://github.com/dotnet/aspire/tree/davidfowl/bicep-resource/src/Aspire.Hosting.Azure/Bicep

**Comment by bjorkstromm** [comment]
> https://github.com/dotnet/aspire/tree/davidfowl/bicep-resource/src/Aspire.Hosting.Azure/Bicep

Interesting 🤔 So instead of using .NET ARM SDK, you're looking into provisioning using Bicep 🙂

**Comment by davidfowl** [comment]
Yep. We’re hoping to get rid of this package by GA so we can rely on azd for both local dev and deploy scenarios.

the manifest modeling let’s azd get of the business of defining every resource type. We just need a few primitives.

**Comment by davidfowl** [comment]
@bjorkstromm let me know what you want to do. I service bus is the second last resource I need to migrate. Might be better to do a follow up on my PR (which will have crappy bicep) after I'm done.

**Comment by bjorkstromm** [comment]
@davidfowl yes, I can do a follow up PR. Don't need to wait for me 🙂

**Comment by bjorkstromm** [comment]
Looking at this now, and it seems we'll run into a use-case not yet solved by the Bicep provisioner. 

Since Bicep lacks support for nested loops I can't do something like this
```bicep
param queues array = []
param topics object = {}

resource serviceBusNamespace 'Microsoft.ServiceBus/namespaces@2022-10-01-preview' = {
  ...

  resource queue 'queues@2022-10-01-preview' = [for name in queues:{
    name: name
  }]

  resource topic 'topics@2022-10-01-preview' = [for topic in items(topics):{
    name: topic.key
    
    resource subscription 'subscriptions@2022-10-01-preview' = [for subscription in topic.value.subscriptions: {
      name: subscription.name
    }]
  }]
}
```

And instead I'd need a module for the topics (see https://stackoverflow.com/a/76105656).

To solve this, should I 
a) Make `AzureBicepServiceBusTopic` (and maybe Queue also) inherit from `AzureBicepResource` and create a separate Bicep files. Haven't looked into the Bicep provisioner, but hopefully I could have the topic resource to depend on the Service Bus namespace?
b) Add support for dependent modules in `BicepTemplateFile` and make `AzureBicepResource.GetBicepTemplateFile` emit all needed files to disk upon provisioning.
c) Try to get everything working in a single Bicep file. Might not even be doable, and it sure won't be pretty :smile:

Anyway, opened draft PR here #2232 where we can discuss further.

**Comment by davidfowl** [comment]
b) @mitchdenny predicted that we would need to do this, so it seems like a good thing to do. 

d) Use a template string and generate the bicep in C# instead of using a resource file.

**Comment by bjorkstromm** [comment]
> b) @mitchdenny predicted that we would need to do this, so it seems like a good thing to do. 
> 
> d) Use a template string and generate the bicep in C# instead of using a resource file.

Thanks! I'll go with option b), and hopefully get the PR ready for review in the next days.

**Comment by davidfowl** [comment]
> Thanks! I'll go with option b), and hopefully get the PR ready for review in the next days.

This is a big change 😄, full transparency, there's a big PR here in azd with this support https://github.com/Azure/azure-dev/pull/3349 that we've been testing. You can test your multi file support with it and see if we need to make more changes there (but I hope we don't).

@bjorkstromm We had a change WRT child resource and how they appear in the manifest. 

1. It needs to be referenced by other resources as a value/connection string
2. It needs to be provisioned by the deployment tool.

Topics and queues are currently provisioned by the service bus (well namespace) bicep file so there's no need for them to appear in the manifest at all outside of being parameters.

Just an FYI.

Thanks again!

**Comment by davidfowl** [comment]
Also we're going to be deleting the old service bus resource as part of this so expect a little churn.

**Comment by bjorkstromm** [comment]
> > Thanks! I'll go with option b), and hopefully get the PR ready for review in the next days.
> 
> This is a big change 😄, full transparency, there's a big PR here in azd with this support [Azure/azure-dev#3349](https://github.com/Azure/azure-dev/pull/3349) that we've been testing. You can test your multi file support with it and see if we need to make more changes there (but I hope we don't).
> 

Does this mean that the `BicepProvisioner` will be removed / or just shell out to `azd` instead of doing the ARM depoyment? Or is the linked PR "just" for azd deployment using manifest? Do you see that we still need `BicepProvisioner` that understands `AzureBicepResource` and can provision with the "F5 experience"?

**Comment by davidfowl** [comment]
Fixed!

**Comment by davidfowl** [comment]
> Does this mean that the BicepProvisioner will be removed / or just shell out to azd instead of doing the ARM depoyment? Or is the linked PR "just" for azd deployment using manifest? Do you see that we still need BicepProvisioner that understands AzureBicepResource and can provision with the "F5 experience"?

Long term we want to just use azd for this flow as well https://github.com/dotnet/aspire/issues/1758. We'll keep the provisioner until we make that transition.
